crates/rate_limit/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+use {
+    chrono::{DateTime, Duration, Utc},
+    deadpool_redis::{Pool, PoolError},
+    moka::future::Cache,
+    redis::{RedisError, Script},
+    std::{collections::HashMap, sync::Arc},
+};
+
+#[derive(Debug, thiserror::Error)]
+#[error("Rate limit exceeded. Try again at {reset}")]
+pub struct RateLimitExceeded {
+    reset: u64,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InternalRateLimitError {
+    #[error("Redis pool error {0}")]
+    Pool(PoolError),
+
+    #[error("Redis error: {0}")]
+    Redis(RedisError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RateLimitError {
+    #[error(transparent)]
+    RateLimitExceeded(RateLimitExceeded),
+
+    #[error("Internal error: {0}")]
+    Internal(InternalRateLimitError),
+}
+
+/// Rate limit check using a token bucket algorithm for one key and in-memory
+/// cache for rate-limited keys. `mem_cache` TTL must be set to the same value
+/// as the refill interval.
+pub async fn token_bucket(
```

Comment:
> `mem_cache` TTL must be set to the same value as the refill interval.

What does this mean? Does `Cache` support different keys with different refill intervals e.g. for what Notify Server needs?